lib/ordered_set.h

```diff
@@ -17,234 +17,540 @@ limitations under the License.
 #ifndef _LIB_ORDERED_SET_H_
 #define _LIB_ORDERED_SET_H_
 
+#include <boost/container/set.hpp>
```

Comment:
Why use `boost::set` rather than `std::set` -- the latter is more portable and reliable

Reply:
Please accept apologies for my replying late.

As explained in my first comment above, the Boost version of `set` attempts to merge the colour flag with one of the pointers => each node in the tree demands smaller block on the heap.

Reply:
We are in the processing of getting rid of boost in the compiler. On the other hand, we now have support for Abseil.